### PR TITLE
INFINITY-2189 Resolve StateStore.storeStatus() TODO

### PR DESCRIPTION
**Pull request by nickbp**
... and add unit tests for task id repair.

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 1]
```diff
@@ -201,7 +251,7 @@ public static void storeTaskStatusAsProperty(StateStore stateStore, String taskN
      */
     public static void setLastCompletedUpdateType(
             StateStore stateStore,
-            ConfigurationUpdater.UpdateResult.DeploymentType updateResultDeploymentType) {
+             ConfigurationUpdater.UpdateResult.DeploymentType updateResultDeploymentType) {
```
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: extra space at start of line

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java [hunk 2]
```diff
@@ -154,7 +157,62 @@ public void testStateStoreWithMismatchedIdRaisesErrorOnStatus() {
         TaskStatus taskStatus = newTaskStatus(CommonIdUtils.toTaskId("not-" + taskName), TaskState.TASK_UNKNOWN);
 
         assertThat(taskInfo.getTaskId(), is(not(taskStatus.getTaskId())));
-        StateStoreUtils.getTaskInfo(stateStore, taskStatus);
+        StateStoreUtils.getTaskName(stateStore, taskStatus);
+    }
+
```
Comment (gabrielhartmann) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: Why double newlines?  Been writing in python recently?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 3]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
         }
 
-        return taskInfoOptional.get();
+        return taskInfoOptional.get().getName();
+    }
+
+    /**
+     * TaskInfo and TaskStatus objects referring to the same Task name are not written atomically.
+     * It is therefore possible for the states across these elements to become out of sync.  While the scheduler process
+     * is up they remain in sync.  This method guarantees produces an initial synchronized state.
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: "guarantees produces": choose one.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 4]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
+     * <ol>
+     * <li>TaskInfo(name=foo, id=1) is written</li>
+     * <li>TaskStatus(name=foo, id=1, status=RUNNING) is written</li>
+     * <li>Task foo is reconfigured/relaunched</li>
+     * <li>TaskInfo(name=foo, id=2) is written</li>
+     * <li>Scheduler is restarted before new TaskStatus is written</li>
+     * <li>Scheduler comes back and sees TaskInfo(name=foo, id=2) and TaskStatus(name=foo, id=1, status=RUNNING)</li>
+     * </ol>
+     *
+     * Note that this ID mismatch would specifically cause problems {@link #getTaskName(StateStore, TaskStatus)} which
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: cause problems **in**

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 5]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
+     * requires that the ids align.
+     */
+    static void repairTaskIDs(StateStore stateStore) {
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Does this method guarantee synchronous  access? That is to say, is there another method that may be updating the `stateStore` at the same time as this one being called?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 6]
```diff
@@ -779,7 +779,9 @@ public void statusUpdate(SchedulerDriver driver, Protos.TaskStatus status) {
 
             // Store status, then pass status to PlanManager => Plan => Steps
             try {
-                stateStore.storeStatus(status);
+                String taskName = StateStoreUtils.getTaskName(stateStore, status);
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: @benclarkwood mentioned that we prefer `final` for method-local variables.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 7]
```diff
@@ -799,8 +801,7 @@ public void statusUpdate(SchedulerDriver driver, Protos.TaskStatus status) {
                     // Map the TaskStatus to a TaskInfo. The map will throw a StateStoreException if no such
                     // TaskInfo exists.
                     try {
-                        Protos.TaskInfo taskInfo = StateStoreUtils.getTaskInfo(stateStore, status);
-                        StateStoreUtils.storeTaskStatusAsProperty(stateStore, taskInfo.getName(), status);
+                        StateStoreUtils.storeTaskStatusAsProperty(stateStore, taskName, status);
                     } catch (StateStoreException e) {
                         LOGGER.warn("Unable to store network info for status update: " + status, e);
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: Would it make sense to log `taskName` here as well or does the output of `status` provide sufficient context?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStore.java [hunk 8]
```diff
@@ -202,6 +193,7 @@ public void storeStatus(Protos.TaskStatus status) throws StateStoreException {
         } catch (PersisterException e) {
             throw new StateStoreException(e);
         }
+        return currentStatusOptional;
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStore.java]: The naming of `currentStatusOptional` is a little misleading now that we return it. The comments also mention the "previous `TaskStatus`".

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 9]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
+        for (Protos.TaskInfo task : stateStore.fetchTasks()) {
+            Optional<Protos.TaskStatus> statusOptional = stateStore.fetchStatus(task.getName());
+
+            if (statusOptional.isPresent()) {
+                Protos.TaskStatus status = statusOptional.get();
+                if (!status.getTaskId().equals(task.getTaskId())) {
+                    LOGGER.warn("Found StateStore status inconsistency: task.taskId={}, taskStatus.taskId={}",
+                            task.getTaskId(), status.getTaskId());
+                    repairedTasks.add(task.toBuilder().setTaskId(status.getTaskId()).build());
+                    if (!status.getTaskId().getValue().equals("")) {
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Here we explicitly check for `status.getTaskId() == ""`, but on line 155, we may be using this status to update the ID of the task. Does this not cause further inconsistency?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 10]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
+                        repairedStatuses.put(task.getName(),
+                                status.toBuilder().setState(Protos.TaskState.TASK_FAILED).build());
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Should we add a message as we do for the non-present status case?

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreTest.java [hunk 11]
```diff
@@ -20,7 +20,7 @@
 /**
  * Tests to validate the operation of the {@link StateStore}.
  */
-public class DefaultStateStoreTest {
+public class StateStoreTest {
     private static final Protos.FrameworkID FRAMEWORK_ID =
             Protos.FrameworkID.newBuilder().setValue("test-framework-id").build();
     private static final String TASK_NAME = "test-task-name";
```
Comment (elezar) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreTest.java]: Isn't this `TASK_NAME` a duplicate of `TestConstants.TASK_NAME`? What is the motivation for using static members with the same names from `TestConstants` in general?

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java [hunk 12]
```diff
@@ -154,7 +157,62 @@ public void testStateStoreWithMismatchedIdRaisesErrorOnStatus() {
...
+
+    @Test
+    public void testRepairTaskIdsMissingStatus() {
```
Comment (elezar) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: Does it make sense to test for the happy case?

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java (hunk 2 above)
Comment (elezar) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: @nickbp @gabrielhartmann @benclarkwood can we share our code style with each other and set format on save?

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java [hunk 13]
```diff
@@ -154,7 +157,62 @@ public void testStateStoreWithMismatchedIdRaisesErrorOnStatus() {
...
+        TaskInfo taskInfo = newTaskInfo(taskName);
+        stateStore.storeTasks(ImmutableList.of(taskInfo));
+        TaskID taskID = TaskID.newBuilder().setValue("").build();
+        TaskStatus taskStatus = newTaskStatus(taskID, TaskState.TASK_UNKNOWN);
+        stateStore.storeStatus(taskName, taskStatus);
+
+        StateStoreUtils.repairTaskIDs(stateStore);
+
+        // TaskInfo was updated to match Status' id:
+        assertThat(stateStore.fetchTask(taskName).get().getTaskId(), is(taskID));
```
Comment (elezar) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: Just to be clear, this means that both IDs are now empty?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 14]
```diff
@@ -121,7 +119,59 @@ public static TaskInfo getTaskInfo(StateStore stateStore, TaskStatus taskStatus)
...
+    }
+
+    /**
+     * TaskInfo and TaskStatus objects referring to the same Task name are not written atomically.
+     * It is therefore possible for the states across these elements to become out of sync.  While the scheduler process
+     * is up they remain in sync.  This method guarantees produces an initial synchronized state.
+     *
+     * For example:
+     * <ol>
+     * <li>TaskInfo(name=foo, id=1) is written</li>
```
Comment (elezar) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Does the implementation make assumptions about ordering? Is it possible to have a `TaskStatus` written before a `TaskInfo`, for example?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java (hunk 6 above)
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: Let me re-phrase my earlier comment:

`final` is a good practice for method-local variables. We do not have any existing preference or enforcement of this. If we want to add it in as a static analysis check then we can do so.

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java (hunk 12 above)
Comment (nickbp) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: Sure, done.

**Comment by NimaVaziri** [comment]
@nickbp looks like this should be good to merge upon rebase + conflict resolution.

**Pull request merged**